Integrations getting started guide (#4314)
DataDog/documentation: content/en/getting_started/integrations.md

---
title: Getting started with Integrations
kind: documentation
aliases:
  - /getting-started-integrations/
  - /guides/getting-started-integrations/
  - /guides/integrations/
further_reading:
- link: "https://learn.datadoghq.com/course/view.php?id=2"
  tag: "Learning Center"
  text: "Introduction to Datadog"
- link: "/integrations/"
  tag: "Integrations"

>   text: "Our full list of integrations"

Reviewer: ```suggestion
  text: "Datadog's full list of integrations"
```

- link: "/graphing/"
  tag: "Graphing"

>   text: "Now that you have your integrations set up, build graphs to visulize your data"

Reviewer: ```suggestion
  text: "Now that you have your integrations set up, build graphs to visualize your data"
```

- link: "/monitors/"
  tag: "Alerts"
  text: "Set up alerts on your data"
---

> An integration at the highest level is when you assemble a unified system from units that are usually considered separately. At Datadog, you can use integrations to bring together all of the units of your infrastructure and gain insight into the unified system as a whole - you can see pieces individually and also how individual pieces are impacting the whole. Datadogs integrations use a Python class called a check to define the typical metrics most people would want to monitor for that system.

Reviewer: ```suggestion
An integration, at the highest level, is when you assemble a unified system from units that are usually considered separately. At Datadog, you can use integrations to bring together all of the units of your infrastructure and gain insight into the unified system as a whole—you can see pieces individually and also how individual pieces are impacting the whole. Datadog's integrations use a Python class called `check` to define the typical metrics most people would want to monitor for that system.
```

I'm not sure now necessary it is to explain integrations at such a high level. I think even our least technical users are still a bit technical enough to know what "integration" means in this context. It is kind of nice to have the explanation there, though... Things to think about.

Author: Yeah, this is an interesting point to think about. I didn't spend too much time on it, but Kirk told me that this is actually a question that comes up a lot at events.

> * Over 250 integrations are [officially listed][1], and supported in our [integrations core][2] (and we're always adding more).

Reviewer: ```suggestion
* Over 250 integrations are [officially listed][1], and supported in the [integrations core][2] (and more are always being added).
```
I think the guides can take on a more relaxed tone than the rest of the docs, but we still shouldn't use we/our

* Even more integrations have been built by our open source community, in [integrations extras][3].
* You can build your own integration using [Datadog API][4].
* The Agent is [open source][5] and you can instrument your own if you'd like.
* All incoming data is treated the same throughout Datadog.

> Whether you're beginning developing a new product today or you've been developing for decades, it's never too early or too late to start monitoring with Datadog.

Reviewer: ```suggestion
Whether you're beginning development on a new project today, or you've been developing for decades, it's never too early or too late to start monitoring.
```

changing from "beginning developing" to "beginning development on" because i don't like gerund-gerund it's just weird-sounding

changed product to project, since not everyone uses DD for products; some of them are just using it for hobbies/funsies

removed "with datadog"—with that at the end, the whole sentence comes off as a bit marketingy, but it is true that it's never too early or too late to start monitoring! also having a transitional sentence is nice.

Author: Cool. I am trying to answer the question here of "when should I start using this?" I think the answer to that is as early as possible is better, but that being said, you can really start at any time. I think it's important information, but I am not sure how to communicate that without being marketingy

### Key terms

**parameters** - Use the parameters in the `conf.yaml` file to control accesses between your integration data source and the Agent. The individual integrations `conf.yaml.example` file has all of the required and not required parameters listed. Define all of the needed parameters for all of your integrations in a single `conf.yaml` file - if you have more than one file, the Agent only uses the alphabetically first file.

> **conf.yaml** - You can find the `conf.yaml` in the `conf.d/` folder at the root of your [Agent's configuration directory][6]. Use this file to connect and configure any integrations to your system. Non required parameters are usually commented out.

Reviewer: ```suggestion
**conf.yaml** - You can find the `conf.yaml` in the `conf.d/` folder at the root of your [Agent's configuration directory][6]. Use this file to connect integrations to your system, as well as configure their settings. Nonrequired parameters are usually commented out.
```

### Tagging

Tagging is a key part of filtering and aggregating the data coming into Datadog across many sources. You can assign tags in configuration files, in environment variables, in the UI, or the API, in DogStatsD, and new integrations automatically inherited existing tags. For more information about tagging, see [Getting started with tags][12].

## Installing multiple integrations

Installing more than one integration is a matter of adding the configuration information to your already existing `conf.yaml` file. Look up the required parameters for the new integration, and add it into your already existing `conf.yaml` file, and then follow the same steps to validate your configuration.

## Security practices

For information on how Datadog handles your data, and other security considerations, see our [Security documentation][21].

## What's next?

Now that you have your first integrations set up, you can start [exploring all of the metrics][22] being sent by Datadog to your application, and use these metrics to begin setting up [graphs][23] and [alerts][24] to monitor your data.

Also check out our [Logs management][13], [APM][25], and [Synthetics][26] solutions.

> ## Troubleshooting

Reviewer: doesn't seem to be anything in this section

Author: Yeah, I was going to ask folks about this. What are the issues that people have that are common to any integration?

Reviewer 2: Troubleshooting and integration validation process are quite similar.. Maybe we could just tell them to check if yaml file are valid, and then redirect to: https://docs.datadoghq.com/agent/troubleshooting/?tab=agentv6

Reviewer 3: ...the most common problem with integration setup is YAML file validation. If you YAML file is invalid an error message displays when checking the Agent's status...

---

> title: Getting started with Integrations

Reviewer 3: ```suggestion
title: Introduction to Integrations
```

kind: documentation
aliases:
  - /getting-started-integrations/
  - /guides/getting-started-integrations/

>   - /guides/integrations/

Reviewer 3: Remove aliases. No need to create aliases for URLs that don't currently exist.

further_reading:
- link: "https://learn.datadoghq.com/"
  tag: "Learning Center"
  text: "Introduction to Datadog"
- link: "/integrations/"
  tag: "Integrations"
  text: "Datadog's full list of integrations"
- link: "/graphing/"
  tag: "Graphing"
  text: "Now that you have your integrations set up, build graphs to visualize your data"
- link: "/monitors/"
  tag: "Alerts"
  text: "Set up alerts on your data"

> ---

Reviewer 3: I'd remove the last 2 further reading links since they aren't related to integrations.

This is a guide for using integrations, for information about building a new integration, see the [Create a new integration][1] page.

> An integration, at the highest level, is when you assemble a unified system from units that are usually considered separately. At Datadog, you can use integrations to bring together all of the units of your infrastructure and gain insight into the unified system as a whole—you can see pieces individually and also how individual pieces are impacting the whole. Datadog's integrations use a Python class called `check` to define the typical metrics most people would want to monitor for that system.

Reviewer 3: `Datadog's integrations use a Python class called check to define the typical metrics most people would want to monitor for that system.`

This is true of Agent based integrations but not authentication (crawler) based integrations.

> An integration, at the highest level, is when you assemble a unified system from units that are usually considered separately. At Datadog, you can use integrations to bring together all of the units of your infrastructure and gain insight into the unified system as a whole—you can see pieces individually and also how individual pieces are impacting the whole. Datadog's integrations use a Python class called `check` to define the typical metrics most people would want to monitor for that system.

Reviewer 3: ```suggestion
An integration, at the highest level, is when you assemble a unified system from units that are usually considered separately. At Datadog, you can use integrations to bring together all of the units of your infrastructure and gain insight into the unified system as a whole—you can see pieces individually and also how individual pieces are impacting the whole.

Datadog provides two main types of integrations:

* Agent based integrations are installed with the Datadog Agent and use a Python class called `check` to define the metrics to collect.
* Authentication (crawler) based integrations are set up in the [Datadog App](https://app.datadoghq.com/account/settings) where you provide credentials for obtaining metrics via API.
```

Put link in link list 😄

> * Over 250 integrations are [officially listed][2], and supported in the [integrations core][3] (and more are always being added).

Reviewer 3: ```suggestion
* Over 250+ integrations are [officially listed][2], and supported in the [integrations core][3].
```

Reviewer 2: The way i understand this sentence is that all integrations are living in integrations core, which is not true. 
We have indeed more than 250 integrations. All Agent related integrations officially supported by Datadog are in integrations-core, All Agent related integrations supported by the Datadog Community are in integrations-extras and all the other can be found in the DD integrations page: https://app.datadoghq.com/account/settings

* Even more integrations have been built by our open source community, in [integrations extras][4].

> * You can build your own integration using [Datadog API][5].

Reviewer 3: ```suggestion
* You can build your own integration using the [Datadog API][5].
```

> * The Agent is [open source][6] and you can instrument your own if you'd like.

Reviewer 3: Instrument your own what? If it's `integration` isn't this covered under `You can build your own`?

Reviewer 2: Yup, we should tell the user that they can create their own custom check: https://docs.datadoghq.com/developers/write_agent_check/?tab=agentv6

> * All incoming data is treated the same throughout Datadog.

Reviewer 3: Not sure of the need for this bullet point?

Reviewer 2: +1

> It's best to start monitoring your projects as early in the development process as possible, but you can start at any stage.

Reviewer 3: ```suggestion
It's best to start collecting metrics for your projects as early in the development process as possible, but you can start at any stage.
```

I think we should keep the word `monitoring` for the `monitoring` section because it could get confusing. Installing / setting up an integration is about collecting the metrics. Monitoring is about watching the metrics.

> ### Key terms

Reviewer 2: We should also talk about the `events` and the `service checks` in this section.

**parameters** - Use the parameters in the `conf.yaml` file to control accesses between your integration data source and the Agent. The individual integrations `conf.yaml.example` file has all of the required and not required parameters listed. Define all of the needed parameters for all of your integrations in a single `conf.yaml` file - if you have more than one file, the Agent only uses the alphabetically first file.

> **conf.yaml** - You can find the `conf.yaml` in the `conf.d/` folder at the root of your [Agent's configuration directory][7]. Use this file to connect integrations to your system, as well as configure their settings. Nonrequired parameters are usually commented out.

Reviewer 2: Prior to any configuration there are no conf.yaml files, there are only conf.yaml.example files.

> **instance** - You define and map the instance of whatever you are monitoring in the `conf.yaml` file. For example, in the [`http_check` integration][8], you're defining the name associated with the instance of the HTTP endpoint you are monitoring up and downtime. You can monitor multiple instances in the same integration, and you do that by defining all of the instances in the `conf.yaml` file.

Reviewer 2: We could add a link to the multiple instance example here. An instance is an object defined under the `instances:` parameter:

```
instances:

  # this is my first instance
  - url: example.com

  # this is my second instance
  - url: example_2.com
```

**metadata.csv** - The file that lists and stores the metrics collected by each integration.

**custom check** - If you have a unique system that you want to monitor, or if you're going to expand the metrics already sent by an integration, you can build a [custom check][11] to define and send metrics to Datadog. However, if you want to monitor a generally available application, public service, or an open source project, and we don't have an integration that already exists for it, you should consider [building a new integration][1] instead of a custom check.

**tagging** - [Tags][12] are a way to add customization to metrics so that you can filter and visualize them in the most useful way to you.

**logging** - If the system you are monitoring has logs, you can customize the logs you are sending to Datadog and use our [Logging Management solution][13] to manage and analyze them.

## Setting up an integration

> The Datadog Agent package includes most integrations, so you don't need to download each individually.

Reviewer 2: This sentence is a little bit too fuzzy, Datadog Agent package includes integrations officially supported by Datadog, they can be found here: https://github.com/DataDog/integrations-core, community based integration can also be used but are not packaged with the Agent, for more information on how to install them or how to manage officialy supported integrations they can refer to this guide: https://docs.datadoghq.com/agent/guide/integration-management/

### API and Application keys

> In order to download the the Datadog Agent package, you need an [API and Application key][14]. You can manage your accounts API and Application keys in the [API Settings page][15] of the UI.

Reviewer 2: We could add a link to there: https://app.datadoghq.com/account/settings#agent for the "download agent"

> In order to download the the Datadog Agent package, you need an [API and Application key][14]. You can manage your accounts API and Application keys in the [API Settings page][15] of the UI.

Reviewer 3: ```suggestion
In order to install the the Datadog Agent package, you need an [API and application key][14]. You can manage your accounts API and application keys in the [API Settings page][15] of the UI.
```

### Installation

If you want to connect with a cloud service provider, navigate to that provider on the [Integrations page][16] for specific instructions on how to connect. For most other integrations, start by installing the [Datadog Agent][17]. Most integrations are supported on our containerized agents: [Docker][18], and [Kubernetes][19]. After you've downloaded the Agent, go to the [Integrations page][16] section to find specific configuration instructions for individual integrations.

> ### Configuration

Reviewer 2: We should split this section in Two, Configuration instructions for cloud service providers that can be found: https://app.datadoghq.com/account/settings#integrations

And configuration for Agent integrations

> Configurations are more specifc to [individual integrations][16]. Configure most integrations by editing the `conf.yaml` file, in the `conf.d/` folder at the root of your Agent's configuration directory. Most integration have a sample `conf.yaml` that lists all available configuration options.

Reviewer 2: In the conf.d folder, there are one folder per integrations named `<INTEGRATIONS>.d` in each integration folder there is a `conf.yaml.example` file that act as an example file. First step is to rename this file to `conf.yaml` in order to activate the integration. Then inside the configuration file some parameters are required to be set-up some are optional. All configuration files follow this norm: https://docs.datadoghq.com/developers/integrations/new_check_howto/#parameters-documentation

> For example, this is a sample `conf.yaml` for the `http_check` integration:

Reviewer 2: Http check might not be the best fit since it's not a real "integration" in a sense that it doesn't actually connect to a given technology. Maybe using something more common and easy to setup like `apache` might be better.

Reviewer 2: The idea is to explain the required and optional parameter logic.

Reviewer 2: We should also explain in this example the Log collection logic at the bottom of the file.

Reviewer 2: We should also explain how to set-up multiple instances in the same check configuration file.. If the first instance is defined for the local host, we could define the second one on a remote endpoint (this killer feature is often under-estimated by customers.

```
init_config:

instances:
  - name: Example website
    url: https://example.com/
    # disable_ssl_validation: false      # default is true, so set false to check SSL validation
    # ca_certs: /path/to/ca/file         # e.g. /etc/ssl/certs/ca-certificates.crt
    # check_certificate_expiration: true # default is true
    # days_warning: 28                   # default 14
    # days_critical: 14                  # default 7
    # timeout: 3                         # in seconds. Default is 10.
  - name: Example website (staging)
    url: http://staging.example.com/
```

### Validation

To validate your configuration, [run the Agent's `status` subcommand][20], and look for new configuration under the Checks section.

> ### Tagging

Reviewer 2: Tagging should be before the validation section since validation is the last step of the process. 

In the case of Agent integrations, they can either define tags in the datadog.yaml file, and so those tags are applied to ALL integrations data. Or they can define tags in the corresponding integrations configuration file in order to apply tags to data only emitted by this particular integration.

Tags Environment variables for the Agent will apply tags to all Data. Container labeled with the dd tag label will have only its data tagged.

Tagging is a key part of filtering and aggregating the data coming into Datadog across many sources. You can assign tags in configuration files, in environment variables, in the UI, or the API, in DogStatsD, and new integrations automatically inherited existing tags. For more information about tagging, see [Getting started with tags][12].

## Installing multiple integrations

> 

Reviewer 2: This part should go in the integrations configuration for Agent based integration.

## Setting up an integration

The Datadog Agent package includes integrations officially supported by Datadog, in [integrations core][2]. To use the integrations in integrations core, download the Datadog agent. Community based integrations are in [integrations extras][3], and to use those, you need to download them individually. For more information on installing or managing these integrations, see the [integrations management guide][16].

### API and Application keys

In order to [install the the Datadog Agent package][17], you need an [API and Application key][18]. You can manage your accounts API and Application keys in the [API Settings page][19] of the UI.

### Installation

If you want to connect with a cloud service provider, navigate to that provider on the [Integrations page][20] for specific instructions on how to connect. For other supported integrations, install the [Datadog Agent][21]. Most integrations are supported on our containerized agents: [Docker][22], and [Kubernetes][23]. After you've downloaded the Agent, go to the [Integrations page][20] section to find specific configuration instructions for individual integrations.

### Configuration

> Configurations are specific to [individual integrations][20]. In the `conf.d` folder at the root of your Agent's configuration directory, there is a file named `<INTEGRATIONS>.d` in each integration folder, and a sample `conf.yaml` that lists all available configuration options. 

Reviewer 2: The file is named `conf.yaml.example` and is located inside the `<INTEGRATION>.d` folder

> To begin configuring, rename `<INTEGRATIONS>.d`to `conf.yaml`to activate the integration. Then update the parameters inside the configuration file. Non-required parameters are commented out. All configuration files follow the format documented in the [parameters documentation][24].

Reviewer 2: rename `conf.yaml.example` to (...)

## Security practices

For information on how Datadog handles your data, and other security considerations, see our [Security documentation][26].

## What's next?

Now that you have your first integrations set up, you can start [exploring all of the metrics][27] being sent by Datadog to your application, and use these metrics to begin setting up [graphs][28] and [alerts][29] to monitor your data.

Also check out our [Logs management][15], [APM][30], and [Synthetics][31] solutions.

## Troubleshooting

The first step to troubleshooting an integration is to use a plugin in your code editor or use one of the many online tools to verify that the YAML is valid. The next step is to run through all of the [Agent troubleshooting][32] steps.

If you continue to have problems, reach out to [our awesome Support team][33].

## Further Reading

{{< partial name="whats-next/whats-next.html" >}}
 
[1]: /developers/integrations/new_check_howto
[2]: https://github.com/DataDog/integrations-core
[3]: https://github.com/DataDog/integrations-extras
[4]: https://app.datadoghq.com/account/settings
[5]: /api
[6]: https://github.com/DataDog/dd-agent

> [7]: 

Reviewer 2: missing link

[8]: https://app.datadoghq.com/account/settings
[9]: /agent/guide/agent-configuration-files/?tab=agentv6#agent-configuration-directory
[10]: https://github.com/DataDog/integrations-core/blob/master/http_check/datadog_checks/http_check/data/conf.yaml.example#L13
[11]: /developers/metrics
[12]: /developers/metrics/custom_metrics
[13]: /developers/write_agent_check/?tab=agentv6
[14]: /tagging
[15]: /logs
[16]: agent/guide/integration-management

> [17]: 

Reviewer 2: missing link

---
title: Introduction to Integrations
kind: documentation
further_reading:
- link: "https://learn.datadoghq.com/"
  tag: "Learning Center"
  text: "Introduction to Datadog"
- link: "/integrations/"
  tag: "Integrations"
  text: "Datadog's full list of integrations"
---

This is a guide for using integrations, for information about building a new integration, see the [Create a new integration][1] page.

An integration, at the highest level, is when you assemble a unified system from units that are usually considered separately. At Datadog, you can use integrations to bring together all of the units of your infrastructure and gain insight into the unified system as a whole—you can see pieces individually and also how individual pieces are impacting the whole.

* Over 250 Agent integrations are supported in the [integrations core][2].
* Even more Agent integrations have been built by our open source community, in [integrations extras][3].
* Find all available integrations on the [Integrations page in the app][4].

> * You can build your own integration using the [Datadog API][5].

Reviewer 4: Not sure about this, we already linked the "create an integration" doc in the previous paragraph where I learned how to do this in Python but now this seems the way I'm supposed to create custom integrations (which is not). IMO this should be something on the line "you can build a custom integration following the instruction at [Create a new integration][1] page."

* The Agent is [open source][6] and you can [create your own custom check][7].

> Datadog provides three main types of integrations:

Reviewer 4: I would mention somehow in this paragraph the existence of custom checks, that are similar to an agent-based integration but with substantial differences in terms of complexity and reuse.

* Agent-based integrations are installed with the Datadog Agent and use a Python class called `check` to define the metrics to collect.
* Authentication (crawler) based integrations are set up in the [Datadog App][8] where you provide credentials for obtaining metrics with the API. These include popular integrations like [Slack][9],[AWS][10],[Azure][11], and [PagerDuty][12].

> * Library integrations allow you to monitor applications based on the language they are written in, like [Node.js][13], or [Python][14].

Reviewer 4: we can mention here that library integrations rely on the Datadog API

It's best to start collecting metrics on your projects as early in the development process as possible, but you can start at any stage.

### Key terms

Find below the key terms for Datadog integrations:

| Term         | Description |
| ------------ | ----------- |
|**parameters** | Use the parameters in the `conf.yaml` file to control accesses between your integration data source and the Agent. The individual integrations `conf.yaml.example` file has all of the required and not required parameters listed. Define all of the needed parameters for all of your integrations in a single `conf.yaml` file - if you have more than one file, the Agent only uses the alphabetically first file. |

> |**conf.yaml** | You create the `conf.yaml` in the `conf.d/` folder at the root of your [Agent's configuration directory][15]. Use this file to connect integrations to your system, as well as configure their settings. Nonrequired parameters are usually commented out.|

Reviewer 5: I'd mention the `integration_name.d` folder as well, like https://docs.datadoghq.com/agent/guide/agent-configuration-files/?tab=agentv6#checks-configuration-files-for-agent-6

|**tagging** | [Tags][19] are a way to add customization to metrics so that you can filter and visualize them in the most useful way to you.|
|**logging** | If the system you are monitoring has logs, you can customize the logs you are sending to Datadog and use our [Logging Management solution][20] to manage and analyze them.|
|**event** | Events are informational messages about your system that are consumed by [the events stream][21] so that you can build monitors on them.|
|**service check** | Service checks are a type of monitor used to track the uptime status of the service. For more information, see the [Service checks guide][22].|

## Setting up an integration

The Datadog Agent package includes integrations officially supported by Datadog, in [integrations core][2]. To use the integrations in integrations core, download the Datadog agent. Community-based integrations are in [integrations extras][3], and to use those, you need to download the [developer toolkit][23]. For more information on installing or managing these integrations, see the [integrations management guide][24].

### API and Application keys

> In order to [install the Datadog Agent][6], you need an [API and Application key][25]. You can manage your accounts API and Application keys in the [API Settings page][26] of the UI.

Reviewer 5: The agent only needs the API key.

### Installation

If you want to connect with a crawler or library based integration, navigate to that provider on the [Integrations page][27] for specific instructions on how to connect. For other supported integrations, install the [Datadog Agent][28]. Most integrations are supported on our containerized agents: [Docker][29], and [Kubernetes][30]. After you've downloaded the Agent, go to the [Integrations page][27] section to find specific configuration instructions for individual integrations.

### Configuring Agent integrations

> Configurations are specific to [individual integrations][27]. In the `conf.d` folder at the root of your Agent's configuration directory, there is a folder named `<INTEGRATIONS>.d` for each officially supported Agent integration which contains a sample `conf.yaml.example` that lists all available configuration options for this particular integration.

Reviewer 5: I'd change the template name to `<INTEGRATION_NAME>.d`

1. Rename the `conf.yaml.example` file (in the corresponding `<INTEGRATIONS>.d` folder) to `conf.yaml`. 
2. Update the required parameters inside the newly created configuration file with the values corresponding to your environment.
3. [Restart the Datadog Agent][31]

**Note**: All configuration files follow the format documented in the [parameters documentation][32].

For example, this is the minimum `conf.yaml` configuration file needed to collect metrics and logs from the [apache integration][33]:

```
init_config:

instances:

  - apache_status_url: http://localhost/server-status?auto

logs:
  - type: file
    path: /var/log/apache2/access.log
    source: apache
    sourcecategory: http_web_access
    service: apache
  - type: file
    path: /var/log/apache2/error.log
    source: apache
    sourcecategory: http_web_access
```

To create multiple instances in the same Agent check to monitor two Apache services, create a new instance with a `-` in the `instances:` section:

```
init_config:

instances:

  - apache_status_url: http://localhost/server-status?auto

  - apache_status_url: http://<REMOTE_APACHE_ENDPOINT>/server-status?auto
```

### Tagging

> Tagging is a key part of filtering and aggregating the data coming into Datadog across many sources. You can assign tags in configuration files, in environment variables, in the UI, or the API, and in DogStatsD. If you define tags in the `datadog.yaml` file, the tags are applied to all of your integrations. Once you've defined the tag in `datadog.yaml`, all new integrations inherit it. If you use a tag environment variable, it applies to all integrations. If you define tags in the corresponding integrations configuration file, it only applies to that specific integration. If you use tags in containers, it applies only to that tag. For more information about tagging, see [Getting started with tags][19].

Reviewer 5: `datadog.yaml` hasn't been defined.

Reviewer 5: > If you use tags in containers, it applies only to that tag

it only applies to that container

### Validation

To validate your configuration, [run the Agent's `status` subcommand][34], and look for new configuration under the Checks section.

## Installing multiple integrations

> Installing more than one integration is a matter of adding the configuration information to your already existing `conf.yaml` file. Look up the required parameters for the new integration, and add it into your already existing `conf.yaml` file, and then follow the same steps to validate your configuration.

Reviewer 5: No not true. It's about writing a new `conf.yaml` in a different `<INTEGRATION2_NAME>.d` folder

### API and Application keys

In order to [install the Datadog Agent][4], you need an [API key][25]. To use most additional Datadog functionality besides submitting metrics and events, you need an [application key][25]. You can manage your accounts API and application keys in the [API Settings page][26] of the UI.

### Installation

If you want to connect with a crawler or library based integration, navigate to that provider on the [Integrations page][27] for specific instructions on how to connect. For other supported integrations, install the [Datadog Agent][28]. Most integrations are supported on our containerized agents: [Docker][29], and [Kubernetes][30]. After you've downloaded the Agent, go to the [Integrations page][27] section to find specific configuration instructions for individual integrations.

### Configuring Agent integrations

> Configurations are specific to [individual integrations][27]. In the `conf.d` folder at the root of your Agent's configuration directory, there is a folder named `<INTEGRATIONS_NAME>.d` for each officially supported Agent integration which contains a sample `conf.yaml.example` that lists all available configuration options for this particular integration.

Reviewer 5: ```suggestion
Configurations are specific to [individual integrations][27]. In the `conf.d` folder at the root of your Agent's configuration directory, there is a folder named `<INTEGRATION_NAME>.d` for each officially supported Agent integration which contains a sample `conf.yaml.example` that lists all available configuration options for this particular integration.
```

> Configurations are specific to [individual integrations][27]. In the `conf.d` folder at the root of your Agent's configuration directory, there is a folder named `<INTEGRATIONS_NAME>.d` for each officially supported Agent integration which contains a sample `conf.yaml.example` that lists all available configuration options for this particular integration.

Reviewer 5: ```suggestion
Configurations are specific to [individual integrations][27]. In the `conf.d` folder at the root of your Agent's configuration directory, there is a folder named `<INTEGRATION_NAME>.d` for each officially supported Agent integration which contains a sample `conf.yaml.example` that lists all available configuration options for this particular integration.
```

> Configurations are specific to [individual integrations][27]. In the `conf.d` folder at the root of your Agent's configuration directory, there is a folder named `<INTEGRATIONS_NAME>.d` for each officially supported Agent integration which contains a sample `conf.yaml.example` that lists all available configuration options for this particular integration.

Reviewer 5: ```suggestion
Configurations are specific to [individual integrations][27]. In the `conf.d` folder at the root of your Agent's configuration directory, there is a folder named `<INTEGRATION_NAME>.d` for each officially supported Agent integration which contains a sample `conf.yaml.example` that lists all available configuration options for this particular integration.
```

To activate a given integration:

> 1. Rename the `conf.yaml.example` file (in the corresponding `<INTEGRATIONS>.d` folder) to `conf.yaml`. 

Reviewer 5: ```suggestion
1. Rename the `conf.yaml.example` file (in the corresponding `<INTEGRATION_NAME>.d` folder) to `conf.yaml`. 
```

> 1. Rename the `conf.yaml.example` file (in the corresponding `<INTEGRATIONS>.d` folder) to `conf.yaml`. 

Reviewer 5: ```suggestion
1. Rename the `conf.yaml.example` file (in the corresponding `<INTEGRATION_NAME>.d` folder) to `conf.yaml`. 
```

> 1. Rename the `conf.yaml.example` file (in the corresponding `<INTEGRATIONS>.d` folder) to `conf.yaml`. 

Reviewer 5: ```suggestion
1. Rename the `conf.yaml.example` file (in the corresponding `<INTEGRATION_NAME>.d` folder) to `conf.yaml`. 
```

> 1. Rename the `conf.yaml.example` file (in the corresponding `<INTEGRATIONS>.d` folder) to `conf.yaml`. 

Reviewer 5: ```suggestion
1. Rename the `conf.yaml.example` file (in the corresponding `<INTEGRATION_NAME>.d` folder) to `conf.yaml`. 
```

> 1. Rename the `conf.yaml.example` file (in the corresponding `<INTEGRATIONS>.d` folder) to `conf.yaml`. 

Reviewer 5: ```suggestion
1. Rename the `conf.yaml.example` file (in the corresponding `<INTEGRATION_NAME>.d` folder) to `conf.yaml`. 
```

> 1. Rename the `conf.yaml.example` file (in the corresponding `<INTEGRATIONS>.d` folder) to `conf.yaml`. 

Reviewer 5: ```suggestion
1. Rename the `conf.yaml.example` file (in the corresponding `<INTEGRATION_NAME>.d` folder) to `conf.yaml`. 
```

> 1. Rename the `conf.yaml.example` file (in the corresponding `<INTEGRATIONS>.d` folder) to `conf.yaml`. 

Reviewer 5: ```suggestion
1. Rename the `conf.yaml.example` file (in the corresponding `<INTEGRATION_NAME>.d` folder) to `conf.yaml`. 
```

> 1. Rename the `conf.yaml.example` file (in the corresponding `<INTEGRATIONS>.d` folder) to `conf.yaml`. 

Reviewer 5: ```suggestion
1. Rename the `conf.yaml.example` file (in the corresponding `<INTEGRATION_NAME>.d` folder) to `conf.yaml`. 
```

> 1. Rename the `conf.yaml.example` file (in the corresponding `<INTEGRATIONS>.d` folder) to `conf.yaml`. 

Reviewer 5: ```suggestion
1. Rename the `conf.yaml.example` file (in the corresponding `<INTEGRATION_NAME>.d` folder) to `conf.yaml`. 
```

**metadata.csv** - The file that lists and stores the metrics collected by each integration.

**custom check** - If you have a unique system that you want to monitor, or if you're going to expand the metrics already sent by an integration, you can build a [custom check][7] to define and send metrics to Datadog. However, if you want to monitor a generally available application, public service, or an open source project, and we don't have an integration that already exists for it, you should consider [building a new integration][1] instead of a custom check.

**tagging** - [Tags][13] are a way to add customization to metrics so that you can filter and visualize them in the most useful way to you.

**logging** - If the system you are monitoring has logs, you can customize the logs you are sending to Datadog and use our [Logging Management solution][14] to manage and analyze them.

> **event** - Events are informational messages about your system that are consumed by the events stream.

Reviewer 2: we could add a link to the event stream in app: https://app.datadoghq.com/event/stream

We can also explain that monitors can be built upon events: /monitors/monitor_types/event/

> **service check** - Service checks are a type of monitor used to track the status of the service.

Reviewer 2: We could reference the guide on how to use service check here: https://docs.datadoghq.com/monitors/guide/visualize-your-service-check-in-the-datadog-ui/ :)

## Setting up an integration

The Datadog Agent package includes integrations officially supported by Datadog, in [integrations core][2]. To use the integrations in integrations core, download the Datadog agent. Community-based integrations are in [integrations extras][3], and to use those, you need to download them individually. For more information on installing or managing these integrations, see the [integrations management guide][15].

### API and Application keys

In order to [install the Datadog Agent][6], you need an [API and Application key][16]. You can manage your accounts API and Application keys in the [API Settings page][17] of the UI.

### Installation

If you want to connect with a cloud service provider, navigate to that provider on the [Integrations page][18] for specific instructions on how to connect. For other supported integrations, install the [Datadog Agent][19]. Most integrations are supported on our containerized agents: [Docker][20], and [Kubernetes][21]. After you've downloaded the Agent, go to the [Integrations page][18] section to find specific configuration instructions for individual integrations.

### Configuring Agent integrations

> Configurations are specific to [individual integrations][18]. In the `conf.d` folder at the root of your Agent's configuration directory, there is a file named `<INTEGRATIONS>.d` in each integration folder, and a sample `conf.yaml.example` that lists all available configuration options. 

Reviewer 2: ```suggestion
Configurations are specific to [individual integrations][18]. In the `conf.d` folder at the root of your Agent's configuration directory, there is a folder named `<INTEGRATIONS>.d` for each officially supported Agent integration, and a sample `conf.yaml.example` inside that lists all available configuration options. 
```

> To begin configuring, rename `conf.yaml.example` (in `<INTEGRATIONS>.d`) to `conf.yaml` to activate the integration. Then update the parameters inside the configuration file. Non-required parameters are commented out. All configuration files follow the format documented in the [parameters documentation][22].

Reviewer 2: ```suggestion
To activate a given integration, rename the `conf.yaml.example` file (in the corresponding `<INTEGRATIONS>.d` folder) to `conf.yaml`. Then update the parameters inside the newly created configuration file.
```

> All configuration files have things in common:

Reviewer 2: ```suggestion
**Note**: All configuration files follow the format documented in the [parameters documentation][22]:
```

> * Not required parameters are commented out of the example files.

Reviewer 2: we should replace this bullet list with the quick features of the paramters documentation (the overall format of param_name - type - required/optional ...)

Reviewer 2: We should also add here that some Agent configurations files includes a log collection section at the bottom that can be configured for log collection.

Author: this is the only one where I still think it's better to just link out to the list elsewhere (I did fix the broken link though)

* If the integration supports our Logging solution, you configure it here too.
* You can set up multiple instances in the same file to monitor local host and remote endpoints.

> For example, this is a sample `conf.yaml` for the `apache` integration:

Reviewer 2: ```suggestion
For example, this is the minimum `conf.yaml` configuration file needed to collect metrics and logs from the `apache` integration:
```

  ##
  ## Learn more about tagging: https://docs.datadoghq.com/tagging/
  #
  #  tags:
  #    - <KEY_1>:<VALUE_1>
  #    - <KEY_2>:<VALUE_2>

  ## @param disable_ssl_validation - boolean - optional - default: false
  ## Instructs the check to skip the validation of the SSL certificate of the URL being tested.
  ## Defaults to false, set to true if you want to disable SSL certificate validation.
  #
  #  disable_ssl_validation: false

  ## @param connect_timeout - integer - optional
  ## Overrides the default connection timeout value,
  ## and fails the check if the time to establish the (TCP) connection
  ## exceeds the connect_timeout value (in seconds)
  #
  #  connect_timeout: <VALUE_IN_SECOND>

  ## @param receive_timeout - integer - optional
  ## Overrides the default received timeout value, and fails the check if the time to receive
  ## the server status from the Apache server exceeds the receive_timeout value (in seconds)
  #
  #  receive_timeout: <VALUE_IN_SECOND>

## Log Section (Available for Agent >=6.0)
##
## type - mandatory - Type of log input source (tcp / udp / file / windows_event)

> ## port / path / channel - mandatory - Set port if type is tcp or udp. Set path if type is file. Set channel if type is windows_event

Reviewer 6: I understand why this is all on one line, but when we render it on the docs site it's visually problematic:
![image](https://user-images.githubusercontent.com/625232/55862689-40551900-5b79-11e9-8d29-4bed6122190c.png)

I'm not sure what a solution (if any) is since the standard requires single lines for descriptions. 😕 cc @l0k0ms

Reviewer 2: I think we should crop this example, giving only a reference to the source file at the beginning and give the minimum required configuration for metrics + logs while we remove paramters comments. And instead add some comments like 

`required configuration for metrics collection` and `required configuration for logs collection` or somthing

then create an other example just below where we define two apache check instances.

If you continue to have problems, reach out to [our awesome Support team][31].

## Further Reading

{{< partial name="whats-next/whats-next.html" >}}
 
[1]: /developers/integrations/new_check_howto
[2]: https://github.com/DataDog/integrations-core
[3]: https://github.com/DataDog/integrations-extras
[4]: https://app.datadoghq.com/account/settings
[5]: /api
[6]: https://github.com/DataDog/dd-agent
[7]: /developers/write_agent_check/?tab=agentv6
[8]: https://app.datadoghq.com/account/settings
[9]: /agent/guide/agent-configuration-files/?tab=agentv6#agent-configuration-directory
[10]: https://github.com/DataDog/integrations-core/blob/master/http_check/datadog_checks/http_check/data/conf.yaml.example#L13
[11]: /developers/metrics
[12]: /developers/metrics/custom_metrics
[13]: /tagging
[14]: /logs
[15]: agent/guide/integration-management
[16]: /account_management/faq/api-app-key-management
[17]: https://app.datadoghq.com/account/settings#api
[18]: /integrations
[19]: https://app.datadoghq.com/account/settings#agent
[20]: https://app.datadoghq.com/account/settings#agent/docker
[21]: https://app.datadoghq.com/account/settings#agent/kubernetes

> [22]: /integrations/new_check_howto/#parameters-documentation

Reviewer 2: This link seems broken
